### ISSUE TITLE
File writes can hang when write locks are not cleared up

### ISSUE THREAD

**Issue by alexrs84** [microsoft/vscode#142462]
Issue Type: <b>Bug</b>

If there is File Write latency when saving a file (like remotely over SMB), and you are the type of person that spams Ctrl+S/Cmd+S the file can become stuck in a dirty state forever and can never be saved again until you force quit MS Visual Code.

While a file is stuck in a dirty state, other files in the same workspace/folder can still be opened, edited, and saved as normal. 

Closing a file that is stuck in a dirty state and reopening it within the same Microsoft Visual Code instance will not fix the issue, it will still open in a dirty state and the file can still not be saved - attempting to save results in no activity, no errors. "Revert file" will put the file back into a clean state, but once you've changed it and it becomes dirty again you still cannot save it. While any file is stuck in a dirty state the only way to close the Workspace, or Microsoft Visual Code is to Force Quit.

![Screen Shot 2022-02-07 at 7 48 26 PM](https://user-images.githubusercontent.com/1809932/152902549-692dac42-e6d2-4314-a8b9-77bf6563ad97.png)

**To recreate:**
File System: remote SMB 
File Size: it's confusingly easier to recreate while editing small files versus larger files
File Type: any type, including "plain text"

Repeatedly make a change to and save an open file (like repeatedly tapping Enter to insert a new line while also repeatedly tapping Ctrl+S/Cmd+S - quickly). 

It's easiest for me to recreate by holding down Ctrl or Cmd the entire time, and then quickly tapping S > Enter > S > Enter > S > Enter (.etc).

Eventually, the file will stay dirty in the editor and can never be saved again. **P.S. it can be very difficult to recreate sometimes, but I'm able to recreate while editing a very small (new) file using the mentioned method on two unrelated SMB servers. The amount of times you have to spam a change+save to observe the issue varies greatly - it seems random enough that it could be a network packet loss issue though I'm not observing any packet loss to either of the SMB servers that I've tested this on**

**I cannot recreate the issue with a local file, only a remote(SMB) file.**

This only seems to be the case since the last update. Maybe the changes from https://github.com/microsoft/vscode/commit/19034bc492734fa2c1d34a2051d456e59d3a3951 that aren't yet in my version of 1.64.0 (5554b12acf27056905806867f251c859323ff7e9, 2022-02-03T04:20:17.224Z) will fix the issue?

--

VS Code version: Code 1.64.0 (5554b12acf27056905806867f251c859323ff7e9, 2022-02-03T04:20:17.224Z)
OS version: Darwin x64 21.2.0
Restricted Mode: Yes

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz (12 x 2600)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|1, 2, 2|
|Memory (System)|16.00GB (0.36GB free)|
|Process Argv|--crash-reporter-id 6048ceb8-ffc5-4877-8f7c-cee23ec9bc4a|
|Screen Reader|no|
|VM|0%|
</details>Extensions: none<details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyl392cf:30425750
pythontb:30283811
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
vsc1dst:30433059
pythonvs932:30410667
wslgetstarted:30433507
vscop341:30404997
vsrem710:30416614

```

</details>

**Comment by bpasero** [comment]
This sounds like an issue with the SMB driver to me, if a file cannot be saved we will keep the file dirty until it is saved or reverted. 

Please see if you can find any logged output that could help us find the source of the issue:
* run `code --verbose` from the command line and attach the output
* in Code select **Help | Toggle Developer Tools** and select the **Console**
* attach any output you see in the **Console** you can find

Thanks!

**Comment by alexrs84** [comment]
**First example:**

Save starts failing on line 366
[vscode-app-1644337903865.log](https://github.com/microsoft/vscode/files/8025717/vscode-app-1644337903865.log)

I've attached stdout log as well as it shows file resource lock management: doSave(8) and doSave(9) seem to happen on the same resource lock, after this the file stays stuck in a dirty state. MS Visual Code cannot be closed normally - it must be force quit.

[std-out-vscode-app-1644337903865.log](https://github.com/microsoft/vscode/files/8025797/std-out-vscode-app-1644337903865.log)

**Second example:**
doSave(34) and doSave(35) seem to happen on the same resource lock, with doSave(35) failing "exit - because busy saving file:///Volumes/www/Test.txt" - It stays dirty after this, and MS Visual Code has to be force quit - it will never timeout.

[second-example.log](https://github.com/microsoft/vscode/files/8026211/second-example.log)

Every time I encounter the issue it seems that there is more than one doSave(#) on the same resource lock, rather than the normal pattern I observe else where when things are working correctly: resource-lock-creation -> save() -> resource-lock-dispose.

**Comment by alexrs84** [comment]
Write locks are a great idea, but then shouldn't a Save operation have a failure timeout?

Without a write op timeout, the only recourse for a user is to kill the MS Visual Code process if a file write doesn't finish, which doesn't seem like an unreasonable expectation when dealing with SMB (or similar protocols) - you can't even close folder/workspace in MSVC after that. If it matters, when the file gets stuck dirty in MSVC I don't even see any locks for the file in question in the remote server (smbstatus -l).

Even a long failure timeout would be reasonable.

This issue plagues me so frequently while editing files on a SMB mount in MSVC on macOS Monterey 12.1 (21C52) since the last time MS Visual Code auto-updated for me on Feb 4th. I hope if I can stop my neurotic obsession with saving a file every couple of seconds it can help :)

**Comment by bpasero** [comment]
Yeah I tend to agree, let me look into this for Feb.

Nevertheless, the fact that a write operation is never closed smells like a major underlying issue, no matter what. This would not help at all to understand what is going on.

**Comment by bpasero** [comment]
@alexrs84 I looked at the code again and I am not yet 100% convinced this would be a recent regression. To clarify: yes, I introduced write locks for the underlying file system provider, but truth is that we always had write locking on the level of the file service that sits on top of the file system provider. So I would somewhat assume this issue would have reproduced even in earlier versions.

Can you do this:
* disable automatic updates in settings (`update.mode`), otherwise you would get auto updated to the latest version against
* download VSCode 1.63.2 (before my change): [link](https://az764295.vo.msecnd.net/stable/899d46d82c4c95423fb7e10e68eba52050e30ba3/VSCode-darwin.zip)
* use it with the scenario where you see the issue

Does it reproduce?

**Comment by alexrs84** [comment]
That's actually what I did. On February 9th I reverted to 1.63.2 and disabled auto-updates and I haven't had a single issue since. 

I tried to continue using 1.64.0 on February 8th for one more day after applying some updates to macOS Monterey 12 and restarting, but I ended up having to force quit MSVC around 25+ times and out of frustration I gave up and reverted to 1.63.2 - not a single issue since then and MSVC is pretty much permanently open on that laptop 5 or 6 days a week.

**Comment by bpasero** [comment]
And to confirm, this reproduces when you run with `--disable-extensions`?

**Comment by alexrs84** [comment]
There were absolutely no extensions installed, should I still reinstall 1.64 and try again with the `--disable-extensions` argument?

> VS Code version: Code 1.64.0 (https://github.com/microsoft/vscode/commit/5554b12acf27056905806867f251c859323ff7e9, 2022-02-03T04:20:17.224Z)
> OS version: Darwin x64 21.2.0
> Restricted Mode: Yes
> **Extensions: none**

I also tried to make sure it wasn't because of Git, somehow. In MSVC I opened a single smb:// folder, with a single small plain text file, while no MSVC extensions were installed. The issue still occurred randomly while saving.

---

The issue doesn't seem to happen while using MVSC on Windows.

I wish at least one other person could try saving files over smb:// a moderate number of times over a moderate period of time on macOS Monterey using MSVC 1.64.0 to see if they experience the same occasional (but permanent) hung write locks. 

If there is anything else you'd like me to try, just let me know.

**Comment by bpasero** [comment]
Ok thanks, at this point I would like to selectively disable the resource locking feature to see if it is really the culprit and potentially add some logging. 

Now, I can either do this and provide you with a build or when you are a developer yourself in the area you could follow https://github.com/microsoft/vscode/wiki/How-to-Contribute to setup a locally running VSCode that I can provide a branch for with these changes. What do you think?

**Comment by alexrs84** [comment]
If you can provide me with a branch with the resource locking feature
disabled, that would be perfect.

It might take me a day or two to have time to setup and build locally on
that laptop, but definitely before or on Saturday.

**Comment by bpasero** [comment]
Ok I created the branch `ben/file-lock-tweaks`: https://github.com/microsoft/vscode/tree/ben/file-lock-tweaks

To speed things up, I will produce a real insiders build with my changes for you to test. There are 2 things:
* from a command line set `export VSCODE_DISABLE_WRITE_LOCKS=1` and run `code-insiders` to have resource locking disabled [1]
* in another terminal run normally to see extended logging in the output [2]

**[1]**
To confirm this you will see this output in the "main" logs from the output view

```
[main 2022-02-16T12:14:31.426Z] [Disk FileSystemProvider]: resource locking is disabled (/Users/bpasero/Library/Application Support/code-oss-dev/User/settings.json)
```

**[2]**
If you could reproduce the problem again and attach all logs, I have made the logs more verbose.

Thanks!

**Comment by bpasero** [comment]
@alexrs84 the build is out: https://az764295.vo.msecnd.net/insider/649b2ed3392851d28854266946d38c925ed969ea/VSCode-darwin.zip

You need to disable updates in settings before trying, otherwise insiders will auto update back to the non-patched version.

**Comment by alexrs84** [comment]
Don't know if this helps... but normally, when resource-locking is enabled (during successful saves) I see a lock registered for a moment in the remote Samba service while vscode is saving a file, then deregistered when vscode is finished.

When vscode hangs during a save I don't see any locks registered at all in the remote Samba service (refreshed at a 1 sec interval). 

vscode is permanently busy in a save operation (hanged) at line 793 (02-16 14:19:42):
`while [ 1 ];do printf 'date %(%m-%d %H:%M:%S)T\n' && smbstatus -l;sleep 1;done`
[smb-locks.log](https://github.com/microsoft/vscode/files/8083427/smb-locks.log)

**Comment by bpasero** [comment]
@alexrs84 super helpful, thanks! I need a bit more logging so I started a new build with more logging and will let you know when that is done.

Given it does not reproduce when locking is disabled, it really seems like there is a bug deep down in the node.js based file system provider that creates and disposes locking. That makes diagnosing this a bit easier because that code is not too complex. I have a theory what might go wrong but need more logs to figure out.

I am not entirely sure I understand what goes on in samba share when this error occurs. Just to clarify: the "lock" that VSCode creates for a file is really nothing special and certainly not a real file lock on the underlying file system provider. In other words, this lock is purely in-memory and only for the VSCode process. It is not a file lock we set via some file system APIs. So I would not expect our file locking to trigger anything on the samba share specifically. 
Nevertheless, it is possible that samba itself does some file locking before an attempt is made to save a file. Here is what VSCode does when saving a file in your case:
* `fs.open()` on the provided path to get a file handle
* `fs.write()` all the data into the file handle
* `fs.close()` to release the file handle

This is a bit simplified, there are a few more calls we might do, e.g. a `fdatasync` on `close` to ensure the disk buffer has written to disk.

Anyway, if at all any of these calls is hanging, then this issue would also reproduce without our resource locks. Because the other (older) resource lock that happens on the layer above is still in place and would also deadlock if a `write` operation does not return. 

Will keep you posted when the build is done for trying.

**Comment by bpasero** [comment]
@alexrs84 build is out: https://az764295.vo.msecnd.net/insider/1c4f6423cc4679c2cd00c7e48bb3ea4a1c5946c8/VSCode-darwin.zip

This time, do not bother about the `VSCODE_DISABLE_WRITE_LOCKS` variable, rather try to repro the issue and attach the logs, thanks!

**Comment by alexrs84** [comment]
dev console:
[slower-vscode-app-1645090832302.log](https://github.com/microsoft/vscode/files/8087463/slower-vscode-app-1645090832302.log)

full stdout:
[slower-vscode-insiders-stdout.log](https://github.com/microsoft/vscode/files/8087464/slower-vscode-insiders-stdout.log)

-------------------


Regarding the SMB locks earlier, from what I understand `smbstatus -l` shows all opened resource handles (that includes resources opened with flags: `r`, `r+`, `w`, `w+`) that haven't yet been released - regardless of any oplock/flock, if any.

I notice in `src/vs/platform/files/node/diskFileSystemProvider.ts` that when writing:
- _if windows_: manual truncation, then `fs.open(...)` with "`r+`" flag
- _else_: `fs.open(...)` with `w` flag

So the `w` file handle should show for me in the `smbstatus -l` list as a `RDWR` lock like
```
Locked files:
Pid      DenyMode   Access      R/W        Oplock      SharePath        Name            Time
--------------------------------------------------------------------------------------------------
25136    DENY_NONE  0x20087     RDWR       NONE        /mnt/vo.../www   test/Test.txt   We... 14:19:02 2022
```
and **_should_** remain while vscode is still running in the busy state, unless the previous `doSave()` finished up to and including the `fs.close()` call to release the previous `w` (`RDWR`) handle - so maybe it could help pinpoint what's hanging, like maybe it stops executing after the `fs.close()` but before the internal lock `dispose()`. Probably isn't hanging **during** `fs.open()` or `fs.write()` because the former `RDWR` lock in SMB no longer exists, it was released. 


Anyway, let me know if you need any more `--verbose` logs.

**Comment by bpasero** [comment]
@alexrs84 I found it! I totally missed an issue when implementing this locking and here is what happens:
* resource locks are stored per file handles in a map and a file handle is just a number
* we always have a sequence of `open`, `write(n)`, `close` calls for each file write
* in `open` we store the lock for the handle and in `close` we dispose it

And here is the race condition:
* we call `fs.close(fd)` and then try to remove the lock from the map with that `fd`
* however, after `fs.close(fd)` the operating system is free to assign that exact same `fd` to the next write call for any othe file
* as such, there is a slight chance that our lock for `fd` is being replaced with a new lock after `fs.close(fd)` but before we get a chance to clean up the lock

In code (a bit simplified):

```ts
async close(fd: number): Promise<void> {
    try {
        return await Promises.close(fd);

        // !!!!! The `fd` is free now and can be assigned by the OS for another `write` operation !!!!!!

    } catch (error) {
        throw this.toFileSystemProviderError(error);
    } finally {

        // !!!!! here we wrongly assume that the map for file locks
        // is still associated with our file, but in fact it is the next
        // write operation's file we dispose here
        // As such, the original file lock hangs as never getting 
        // disposed

        const lockForHandle = this.mapHandleToLock.get(fd);
        if (lockForHandle) {
            this.mapHandleToLock.delete(fd);
            lockForHandle.dispose();
        }
    }
}
```

Thanks a lot for the patience and the help 🙏

**Comment by alexrs84** [comment]
Awesome, good job!

Looking forward to being able to turn on auto-updates again when this makes it into a release :)

**Comment by bpasero** [comment]
Let me push a fix and do a build for you to verify 👍

**Comment by bpasero** [comment]
@alexrs84 the build is out: https://az764295.vo.msecnd.net/insider/069a612a1cd220174077b49cbde40953e1d4f32b/VSCode-darwin.zip

If you want to share your e-mail address with me (via `benjpas@microsoft.com`), I can offer some free VSCode stickers for your work here 👍

**Comment by alexrs84** [comment]
@bpasero it's fixed! I can't reproduce with 1.65.0-insider - Commit: 069a612a1cd220174077b49cbde40953e1d4f32b

Awesome thank you, I like stickers!

**Comment by bpasero** [comment]
Thanks!
